File: Programmers/코딩테스트_고득점_Kit/DFS_BFS/level3_여행경로.cpp
// 역시나 솔루션 봄..ㅎ
// https://gyyeom.tistory.com/43
// 중간에 경로 끊긴 경우 대한 처리 부분에 주목
// 백트래킹 어떻게 하는지 주목

#include <algorithm>
#include <string>
#include <vector>

using namespace std;

vector<string> answer;
bool visited[100001];
bool flag = false;
int cnt = 0;

void dfs(string start, vector<vector<string>> tickets) {
  if (cnt == tickets.size()) {
    flag = true;
  }
  answer.push_back(start);
  for (int i = 0; i < tickets.size(); i++) {
    if (!visited[i] &&
        tickets[i][0] == start) {  // 현재 티켓의 시작지가 인자값과 같으면
      visited[i] = true;            // 방문처리
      cnt++;                        // 카운트 증가
      dfs(tickets[i][1], tickets);  // 도착지 기준으로 다시 재귀함수 호출

      if (!flag) {
        answer.pop_back();
        visited[i] = false;  // 백트래킹
      }
    }
  }
}

vector<string> solution(vector<vector<string>> tickets) {
  sort(tickets.begin(), tickets.end());
  dfs("ICN", tickets);
  return answer;
}

File: Programmers/코딩테스트_고득점_Kit/DFS_BFS/level3_여행경로.py
# https://school.programmers.co.kr/learn/courses/30/lessons/43164
# 기본 아이디어는 생각했으나, "경로가 여러 가지"인 것을 처리하는 부분에서 헤맸음
# "모든 티켓을 다 써야 함"
# 알파벳 순서가 앞서는 경로를 먼저 return해야하므로, dfs 전에 sorting해주면 됐었음
# 혼자 풀었었는데, 테케 1,2번이 계속 틀려서 결국 솔루션 봄..ㅎ

from collections import defaultdict

graph = defaultdict(list)
answer = []


def dfs(start):
    while graph[start]:  # 방문할 수 있는 곳 모두 고려
        dfs(graph[start].pop(0))  # 도착지점 중 알파벳 가장 빠른 것으로 다시 출발점 지정

    if not graph[start]:
        answer.append(start)
        return


def solution(tickets):
    for start, end in tickets:
        graph[start].append(end)  # 연결정보 기록

    for start, end in graph.items():
        graph[start].sort()  # 알파벳 순으로 방문해야 하므로 sorting해줌

    dfs("ICN")
    return answer[::-1]  # 순서 반대로 해서 리턴
